docs/DeveloperGuide.md

```diff
@@ -1,25 +1,283 @@
+and optional arguments (`mc/4 ay/2021S1 g/A`).
+
+An optional argument consists of 2 parts, which is delimited by a forward slash. In the example above, there are 3 optional
+arguments, which are `mc/4`, `ay/2021S1` and `g/A`. Each optional argument can be represented in this form: `<key>/<value>`.
+
+In some commands, the optional arguments may be compulsory and is checked by the `Parser` at runtime.
+
+The `parse` method parses the `fullCommand` into these parts before passing them as arguments to `CommandCreator`
+methods and returns a `Command` object with the corresponding arguments.
+
+#### Implementation details
+
+The following sequence diagram shows how the `Parser` works.
+
+![DukeSequenceDiagram](./images/ParserSequenceDiagram.png)
+
+1. The `parse` method of `Parser` is invoked by the calling object. In termiNus, the only object that invokes this
+method is `Duke`. The `fullCommand` is passed an argument, which is the full command entered by the user.
+1. The method parses `fullCommand` into two separate `Strings`, which are `rootCommand` and `commandString`.
+`rootCommand` contains the first word of the command and `commandString` contains the rest of the command with the first
+word removed. This is done using the `split` method of the `String` class, then removing the `rootCommand` from the 
+`fullCommand` before storing it in `commandString`.
+1. The method then invokes the `removeArgumentsFromCommand` method to parse and remove optional arguments from the full
+command. This is done using regular expression parsing which is detailed in the next section. The results are returned
+to the `parse` method and stored in `description`.
+1. The method then invokes the `getArgumentsFromRegex` method to parse the optional arguments from the full command.
+The results are stored in a `HashMap<String, String>`, which is a `HashMap` of key-value pairs, similar to the form of the
+optional argument (`<key>/<value>`). The results are returned to the `parse` method and stored as `argumentsMap`.
+1. The method then checks the `rootCommand` and decides which `Command` to return, which calls `CommandCreator`
+methods with the parsed `argumentsMap`, `description`, and `commandString`.
+1. The results of the `CommandCreator` methods are returned as a `Command` back to the invoker of the `parse` method.
+
+#### Regular expression parsing
+
+Two of the previously mentioned methods, `removeArgumentsFromCommand` and `getArgumentsFromRegex` make use of regular
+expressions to parse the optional arguments.
+
+- The regular expression that parses these optional arguments is `([\w]+/[^\s]+)`. This regular expression matches 1 or more
+alphanumeric characters (denoted by `[\w]+`), followed by a forward slash, then 1 or more of any character except whitespace (denoted by `[^\s]+`).
+- The expression also uses capturing parenthesis to ensure that the parser does not parse the same argument twice.
```

Comment:
Is it possible to provide an example in the form of a picture? It is quite confusing to a first-time reader, given that there are so many different symbols.